src/java.base/share/classes/jdk/internal/util/Preconditions.java

```diff
@@ -36,6 +36,51 @@
  */
 public class Preconditions {
 
+    /**
+     * Utility classes for creating the desired exception which can be used
+     * for {@code outOfBoundsExceptionFormatter}.
+     *
+     * @implNote These inner classes can be syntactically replaced by lambda
+     * expression or method reference, but it's not feasible in practices,
+     * because {@code Preconditions} is used in many fundamental classes such
+     * as {@code java.lang.String}, lambda expressions or method references
+     * exercise many other code at VM startup, this could lead a recursive
+     * calls when fundamental classes is used in lambda expressions or method
+     * references.
+     */
+    private static class StringIndexOutOfBoundsExceptionProducer implements
+                    Function<String, StringIndexOutOfBoundsException> {
+        @Override
+        public StringIndexOutOfBoundsException apply(String s) {
+            return new StringIndexOutOfBoundsException(s);
+        }
+    }
+
+    private static class ArrayIndexOutOfBoundsExceptionProducer implements
+                    Function<String, ArrayIndexOutOfBoundsException> {
+        @Override
+        public ArrayIndexOutOfBoundsException apply(String s) {
+            return new ArrayIndexOutOfBoundsException(s);
+        }
+    }
+
+    private static class IndexOutOfBoundsExceptionProducer implements
+                    Function<String, IndexOutOfBoundsException> {
+        @Override
+        public IndexOutOfBoundsException apply(String s) {
+            return new IndexOutOfBoundsException(s);
+        }
+    }
+
+    public static final BiFunction<String, List<Number>, StringIndexOutOfBoundsException> SIOOBE_FORMATTER
+            = Preconditions.outOfBoundsExceptionFormatter(new StringIndexOutOfBoundsExceptionProducer());
+
+    public static final BiFunction<String, List<Number>, StringIndexOutOfBoundsException> AIOOBE_FORMATTER
```

Comment:
Using incorrect exception type. Suggest you embed as inner class rather than separate declaration, since they are only used in one place.

Reply:
Fixed.

FYI: Current exception message looks like this:
```
Exception in thread "main" java.lang.StringIndexOutOfBoundsException: Range [3, 1) out of bounds for length 6
	at CheckIndex$StringIndexOutOfBoundsExceptionProducer.apply(CheckIndex.java:77)
	at CheckIndex$StringIndexOutOfBoundsExceptionProducer.apply(CheckIndex.java:72)
	at java.base/jdk.internal.util.Preconditions$1.apply(Preconditions.java:159)
	at java.base/jdk.internal.util.Preconditions$1.apply(Preconditions.java:156)
	at java.base/jdk.internal.util.Preconditions.outOfBounds(Preconditions.java:62)
	at java.base/jdk.internal.util.Preconditions.outOfBoundsCheckFromToIndex(Preconditions.java:76)
	at java.base/jdk.internal.util.Preconditions.checkFromToIndex(Preconditions.java:295)
	at CheckIndex.main(CheckIndex.java:110)
```
I think now it expresses more exception information than before(and more consistent).